src/ir/match.h

```diff
@@ -0,0 +1,782 @@
+template<> struct KindTypeRegistry<NumberLitKind> {
+  using matched_t = Literal;
+  using data_t = int32_t;
+};
+template<> struct MatchSelf<NumberLitKind> {
+  bool operator()(Literal& lit, int32_t expected) {
+    return lit.type.isNumber() &&
+           Literal::makeFromInt32(expected, lit.type) == lit;
+  }
+};
+decltype(auto) NumberLit(Literal* binder, int32_t expected) {
+  return Matcher<NumberLitKind>(binder, expected);
+}
+
+// Const
+template<> struct NumComponents<Const*> { static constexpr size_t value = 1; };
+template<> struct GetComponent<Const*, 0> {
+  Literal& operator()(Const* c) { return c->value; }
+};
+template<class S> decltype(auto) ConstMatcher(Const** binder, S s) {
+  return Matcher<Const*, S>(binder, {}, s);
+}
+
+// Unary and AbstractUnary
+struct UnaryK {
+  using Op = UnaryOp;
+  static UnaryOp getOp(Type, Op op) { return op; }
+};
+struct AbstractUnaryK {
+  using Op = Abstract::Op;
+  static UnaryOp getOp(Type type, Abstract::Op op) {
+    return Abstract::getUnary(type, op);
+  }
+};
+template<class T> struct UnaryKind {};
+template<class T> struct KindTypeRegistry<UnaryKind<T>> {
+  using matched_t = Unary*;
+  using data_t = typename T::Op;
+};
+template<class T> struct MatchSelf<UnaryKind<T>> {
+  bool operator()(Unary* curr, typename T::Op op) {
+    return curr->op == T::getOp(curr->value->type, op);
+  }
+};
+template<class T> struct NumComponents<UnaryKind<T>> {
+  static constexpr size_t value = 1;
+};
+template<class T> struct GetComponent<UnaryKind<T>, 0> {
+  Expression*& operator()(Unary* curr) { return curr->value; }
+};
+template<class S> decltype(auto) UnaryMatcher(Unary** binder, UnaryOp op, S s) {
+  return Matcher<UnaryKind<UnaryK>, S>(binder, op, s);
+}
+template<class S>
+decltype(auto) AbstractUnaryMatcher(Unary** binder, Abstract::Op op, S s) {
+  return Matcher<UnaryKind<AbstractUnaryK>, S>(binder, op, s);
+}
+
+// Binary and AbstractBinary
+struct BinaryK {
+  using Op = BinaryOp;
+  static BinaryOp getOp(Type, Op op) { return op; }
+};
+struct AbstractBinaryK {
+  using Op = Abstract::Op;
+  static BinaryOp getOp(Type type, Abstract::Op op) {
+    return Abstract::getBinary(type, op);
+  }
+};
+template<class T> struct BinaryKind {};
+template<class T> struct KindTypeRegistry<BinaryKind<T>> {
+  using matched_t = Binary*;
+  using data_t = typename T::Op;
+};
+template<class T> struct MatchSelf<BinaryKind<T>> {
+  bool operator()(Binary* curr, typename T::Op op) {
+    return curr->op == T::getOp(curr->left->type, op);
+  }
+};
+template<class T> struct NumComponents<BinaryKind<T>> {
+  static constexpr size_t value = 2;
+};
+template<class T> struct GetComponent<BinaryKind<T>, 0> {
+  Expression*& operator()(Binary* curr) { return curr->left; }
+};
+template<class T> struct GetComponent<BinaryKind<T>, 1> {
+  Expression*& operator()(Binary* curr) { return curr->right; }
+};
+template<class S1, class S2>
+decltype(auto) BinaryMatcher(Binary** binder, BinaryOp op, S1 s1, S2 s2) {
```

Comment:
Possibly, although I would hope this all gets inlined away anyhow. It would also have the effect of encouraging matchers to be built all in one expression as intended, rather than built from the previously initialized parts. I will make this change.